# Goal of the project
To clean and analyze a dataset of used vehicles from eBay Kleinanzeigen, a classifieds section of the German eBay website.
The dataset contains 20 columns and 50,000 data points from the original dataset. You can find the dataset [here.](https://data.world/data-society/used-cars-data)

The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which which year the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which which year the car was first registered.
- fuelType - What type of fuel the car uses.
brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

In [1]:
# Import the libraries 
import numpy as np
import pandas as pd

In [2]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

# Clean columns:
- Change the column name from camelCase to snake_case
- Change wording to column names that are easier to understand

In [5]:
autos.rename(columns={
    'yearOfRegistration': 'registration_year',
    'monthOfRegistration': 'registration_month',
    'notRepairedDamage': 'unrepaired_damage',
    'dateCrawled': 'date_crawled',
    'offerType': 'offer_type',
    'abtest': 'ab_test',
    'vehicleType': 'vehicle_type',
    'gearbox': 'gear_box',
    'powerPS': 'power_ps',
    'fuelType': 'fuel_type',
    'dateCreated': 'ad_created',
    'nrOfPictures': 'num_photos',
    'postalCode': 'postal_code',
    'lastSeen': 'last_seen',
                    }, 
            inplace=True
            )
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

# Data exploration and cleaning
Let's explore the data and see columns that need cleaning

In [6]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

- Seller and offer_type columns both have a high number of freq, meaning that all the data in the column should be mostly the same
- Num_photos also looks odd

In [7]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [8]:
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [9]:
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

num_photos seems to have 0s in every column, we'll drop that along with seller and offer type; as they contain mostly identical entries

In [10]:
autos = autos.drop(["offer_type", "num_photos", "seller"], axis=1)

In [11]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price  ab_test vehicle_type  registration_year   gear_box  power_ps  \
0  $5,000  control          bus               2004    manuell       158   
1  $8,500  control    limousine               1997  automatik       286   
2  $8,990     test    limousine               2009    manuell       102   
3  $4,350  control   kleinwagen               2007  automatik        71   
4  $1,350     test        kombi               2003    manuell         0   

    model   odometer  registration_month fuel_type       brand  \
0  andere  150,000km                   3       lpg     peugeot   
1     7er  150,000km                   6    benzin         bmw   
2    golf   70,000km                   7    benzin  volkswagen   
3  fortwo   70,000km                   6    benzin       smart   
4   focus  150,000km                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

- Price column, the dollar sign should be removed
- Odometer column, the KM should be removed
- odometer column name should be odometer_km

In [12]:
autos["price"] = (autos["price"]
                          .str.replace('$', '')
                          .str.replace(',', '')
                          .astype(int)
                 )

In [13]:
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

In [14]:
autos["odometer_km"].describe()

count         50000
unique           13
top       150,000km
freq          32424
Name: odometer_km, dtype: object

In [15]:
autos["odometer_km"] = (autos["odometer_km"]
                             .str.replace("km","")
                             .str.replace(",","")
                             .astype(int)
                             )
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price  ab_test vehicle_type  registration_year   gear_box  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

# Further exploring
Let's have a closer look for data that doesn't look right in
- odometer_km
- price

In [16]:
print(autos["odometer_km"].unique().shape)
print(autos["odometer_km"].describe())
print(autos["odometer_km"].value_counts())

(13,)
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64


odometer column seems to have a greater proportion of higher mileage vehicles than low mileage. Let's explore price next

In [17]:
print(autos["price"].unique().shape)
print(autos["price"].describe())
(autos["price"].value_counts().head(20))

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price, dtype: int64

Some cars have a price that is \$0, let's consider removing those rows. Also, the maximum price of a car is \$100,000,000 , which seems high.

In [18]:
autos["price"].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

In [19]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

There are several items with listing under \\$30 and 1,500 with \\$0. We will also disregard the very high values of listings, as there are listings over \$1,000,000.

Seeing as Ebay is an auction site, it is plausible that vehicles were sold for \$1, however any price above \\$350,000 we will remove due to being unrealistic unlikely.

In [20]:
autos = autos[autos["price"].between(1, 351000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

# Exploring date columns
- date_crawled
- last_seen
- ad_created
- registration_month
- registration_year

We'll investigate these columns further to better understand the listings.

In [21]:
print(autos["date_crawled"].describe())
print('\n')
print(autos["last_seen"].describe())
print('\n')
print(autos["ad_created"].describe())
print('\n')
print(autos["registration_month"].describe())
print('\n')
print(autos["registration_year"].describe())

count                   48565
unique                  46882
top       2016-03-16 21:50:53
freq                        3
Name: date_crawled, dtype: object


count                   48565
unique                  38474
top       2016-04-07 06:17:27
freq                        8
Name: last_seen, dtype: object


count                   48565
unique                     76
top       2016-04-03 00:00:00
freq                     1887
Name: ad_created, dtype: object


count    48565.000000
mean         5.782251
std          3.685595
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max         12.000000
Name: registration_month, dtype: float64


count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


Registration_year's minimum and maximum values do not make sense, as cars were not invented yet in year 1000 and year 9999 has not happened yet.

In [22]:
autos[["date_crawled", "ad_created", "last_seen"]][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [23]:
print(autos['date_crawled']
            .str[:10]
            .value_counts(normalize=True, dropna=False)
            .sort_index()
     )

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64


date_crawled seems to be fairly uniform

In [24]:
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

In [25]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

This column gives us an idea of when the car was last seen, it can act as a good indication of when the car was sold. April 5th, 6th and 7th have a significantly higher percentage than all the other data points in the column; further investigation is required to identify a cause.

# Cleaning registration year
We need to come up with a year range that's acceptable for our needs. Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult, realistically it can be somewhere in the first few decades.

In [26]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [27]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.038793369710697

Listings which their registration year is not between 1900 and 2016 is roughly 4%. We can remove these rows.

In [28]:
autos = autos[autos["registration_year"].between(1900,2016)]

In [29]:
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

Most vehicles were registered in the last 20 years.

# Exploring price by brand

In [30]:
brands = autos["brand"].value_counts(normalize=True)
print(brands.head(10))
brands_top_five = brands[brands > 0.05].index

volkswagen       0.211264
bmw              0.110045
opel             0.107581
mercedes_benz    0.096463
audi             0.086566
ford             0.069900
renault          0.047150
peugeot          0.029841
fiat             0.025642
seat             0.018273
Name: brand, dtype: float64


Volkswagen is by far the most popular brand, it is notable that German brand's make up 4 of the top 5 most popular brands.
Let's take a subset of the most popular brands from our dataset. We'll focus on brands that are greater than 5%.

In [31]:
aggregated_mean_price = {}
for brand in brands_top_five:
    selected_rows = autos[autos["brand"] == brand]
    mean_price = selected_rows["price"].mean()
    aggregated_mean_price[brand] = int(mean_price)

aggregated_mean_price

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

Audi, Mercedes Benz and BMW have the highest mean price of all the vehicles, respectively. This result is not surprising as they are luxury brand vehicles. Opel and Ford are at the bottom of the list in terms of mean price, while Volkswagen is in the middle. Based on Volkswagen's price point, it may provide insight into its popularity.

# Exploring the average mileage

In [32]:
aggregated_mean_mileage = {}
for brand in brands_top_five:
    selected_rows = autos[autos["brand"] == brand]
    mean_mileage = selected_rows["odometer_km"].mean()
    aggregated_mean_mileage[brand] = int(mean_mileage)

Let's convert both aggregated_mean_mileage and aggregated_mean_price into series and then create a dataframe with both.

In [33]:
mileage_series = pd.Series(aggregated_mean_mileage).sort_values(ascending=False)
price_series = pd.Series(aggregated_mean_price).sort_values(ascending=False)

df_brand_info = pd.DataFrame(mileage_series, columns=["mean_mileage"])
df_brand_info

mean_mileage
bmw                  132572
mercedes_benz        130788
opel                 129310
audi                 129157
volkswagen           128707
ford                 124266

In [34]:
df_brand_info["mean_price"] = price_series 
df_brand_info

mean_mileage  mean_price
bmw                  132572        8332
mercedes_benz        130788        8628
opel                 129310        2975
audi                 129157        9336
volkswagen           128707        5402
ford                 124266        3749

The variation in mileage does not vary significantly. There is a slight trend that the more expensive vehicles are priced higher, while the less expensive vehicles having lower mileage.

# Cleaning categorical data with German words: 

In [35]:
autos[["vehicle_type", "gear_box", "fuel_type", "unrepaired_damage"]].head(10)

vehicle_type   gear_box fuel_type unrepaired_damage
0          bus    manuell       lpg              nein
1    limousine  automatik    benzin              nein
2    limousine    manuell    benzin              nein
3   kleinwagen  automatik    benzin              nein
4        kombi    manuell    benzin              nein
5          bus  automatik    diesel               NaN
6    limousine    manuell    benzin               NaN
7    limousine    manuell    diesel              nein
8          NaN    manuell       NaN              nein
9          bus    manuell    benzin              nein

We would like to change these columns to their english counterparts. Let's start with vehicle type

In [36]:
autos["vehicle_type"].unique()

array(['bus', 'limousine', 'kleinwagen', 'kombi', nan, 'coupe', 'suv',
       'cabrio', 'andere'], dtype=object)

In [37]:
autos["vehicle_type"] = autos["vehicle_type"].map({
    'bus': 'bus',
    'limousine': 'limousine',
    'kleinwagen': 'small car',
    'kombi': 'combi',
    'coupe': 'coupe',
    'suv': 'suv',
    'cabrio': 'convertible',
    'andere': 'other'})
#check
autos["vehicle_type"].unique()

array(['bus', 'limousine', 'small car', 'combi', nan, 'coupe', 'suv',
       'convertible', 'other'], dtype=object)

In [38]:
#check
autos["vehicle_type"].unique()

array(['bus', 'limousine', 'small car', 'combi', nan, 'coupe', 'suv',
       'convertible', 'other'], dtype=object)

Next is the gear box column

In [39]:
autos["gear_box"].unique()

array(['manuell', 'automatik', nan], dtype=object)

In [40]:
autos["gear_box"] = autos["gear_box"].map({
    'manuell': 'manual',
    'automatik': 'automatic'})
#check
autos["gear_box"].unique()

array(['manual', 'automatic', nan], dtype=object)

fuel type is next

In [41]:
autos["fuel_type"].unique()

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [42]:
autos["fuel_type"] = autos["fuel_type"].map({
    'lpg': 'lpg',
    'benzin': 'petrol',
    'diesel': 'diesel',
    'cng': 'cng',
    'hybrid': 'hybrid',
    'elektro': 'electric',
    'andere': 'other'})
autos["fuel_type"].unique()

array(['lpg', 'petrol', 'diesel', nan, 'cng', 'hybrid', 'electric',
       'other'], dtype=object)

In [43]:
autos["unrepaired_damage"].unique()

array(['nein', nan, 'ja'], dtype=object)

In [44]:
autos["unrepaired_damage"] = autos["unrepaired_damage"].map({
    'nein': 'no',
    'ja': 'yes'
})
autos["unrepaired_damage"].unique()

array(['no', nan, 'yes'], dtype=object)

# Convert the dates to be uniform numeric data
Might delete this section. Not sure what value it adds or what analysis can be drawn from it.

In [45]:
autos["date_crawled"].describe()

count                   46681
unique                  45124
top       2016-03-19 17:36:18
freq                        3
Name: date_crawled, dtype: object

In [46]:
split_date_crawled = autos["date_crawled"].str.split(" ", expand=True)
split_date_crawled

0         1
0      2016-03-26  17:47:46
1      2016-04-04  13:38:56
2      2016-03-26  18:57:24
3      2016-03-12  16:58:10
4      2016-04-01  14:38:50
...           ...       ...
49995  2016-03-27  14:38:19
49996  2016-03-28  10:50:25
49997  2016-04-02  14:44:48
49998  2016-03-08  19:25:42
49999  2016-03-14  00:42:12

[46681 rows x 2 columns]

Let's drop the time column as it won't be important for our analysis. We'll also convert the date into a numeric

In [47]:
autos["date_crawled"] = split_date_crawled[0].str.replace("-", "")
autos["date_crawled"].astype(int)

0        20160326
1        20160404
2        20160326
3        20160312
4        20160401
           ...   
49995    20160327
49996    20160328
49997    20160402
49998    20160308
49999    20160314
Name: date_crawled, Length: 46681, dtype: int32

In [48]:
autos.head()

date_crawled                                               name  price  \
0     20160326                   Peugeot_807_160_NAVTECH_ON_BOARD   5000   
1     20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   8500   
2     20160326                         Volkswagen_Golf_1.6_United   8990   
3     20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   4350   
4     20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   1350   

   ab_test vehicle_type  registration_year   gear_box  power_ps   model  \
0  control          bus               2004     manual       158  andere   
1  control    limousine               1997  automatic       286     7er   
2     test    limousine               2009     manual       102    golf   
3  control    small car               2007  automatic        71  fortwo   
4     test        combi               2003     manual         0   focus   

   odometer_km  registration_month fuel_type       brand unrepaired_damage  \
0       150000                   3       lpg     peugeot                no   
1       150000                   6    petrol         bmw                no   
2        70000                   7    petrol  volkswagen                no   
3        70000                   6    petrol       smart                no   
4       150000                   7    petrol        ford                no   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

In [49]:
autos["date_crawled"].value_counts(normalize=True)

20160403    0.038774
20160320    0.038024
20160321    0.037317
20160312    0.036824
20160404    0.036610
20160314    0.036332
20160307    0.036246
20160402    0.035561
20160319    0.034661
20160328    0.034597
20160315    0.034361
20160329    0.034104
20160330    0.033804
20160401    0.033804
20160308    0.033547
20160309    0.033247
20160322    0.032840
20160311    0.032454
20160310    0.032240
20160323    0.032197
20160326    0.032069
20160331    0.031790
20160317    0.031790
20160325    0.031512
20160327    0.030783
20160316    0.029498
20160324    0.029477
20160305    0.025192
20160313    0.015874
20160306    0.014160
20160405    0.013003
20160318    0.012810
20160406    0.003085
20160407    0.001414
Name: date_crawled, dtype: float64

# Most common brand/model combination

In [50]:
autos.groupby(["brand", "model"]).size().nlargest(5)

brand       model 
volkswagen  golf      3707
bmw         3er       2615
volkswagen  polo      1609
opel        corsa     1592
volkswagen  passat    1349
dtype: int64

Volwkswagen golf is the most popular brand and model of car. Volkswagen hold three of the top five most popular brand and model in the dataset.

# Exploring odometer km and price
We'll start by identifying the unique values in the column and seeing if we can group them into categories.

In [51]:
autos["odometer_km"].value_counts(normalize=True).sort_index(ascending=False)

150000    0.644481
125000    0.104047
100000    0.044086
90000     0.035839
80000     0.029455
70000     0.025428
60000     0.024164
50000     0.021272
40000     0.017073
30000     0.016281
20000     0.015895
10000     0.005163
5000      0.016816
Name: odometer_km, dtype: float64

I'll split this into 5 groups: Under 30K, under 70K, 80K-100K, above 100K

In [52]:
under_30 = autos[autos["odometer_km"] <= 30000]

under_70 = autos[(autos["odometer_km"] > 30000) & (autos["odometer_km"] <= 70000)]

between80_100 = autos[(autos["odometer_km"] >= 80000) & (autos["odometer_km"] <= 100000)]

above_100 = autos[autos["odometer_km"] > 100000]

In [79]:
#check
print(under_30["odometer_km"].unique())
 
print(under_70["odometer_km"].unique())

print(between80_100["odometer_km"].unique())

print(above_100["odometer_km"].unique())

[10000 30000 20000  5000]
[70000 50000 60000 40000]
[ 80000  90000 100000]
[150000 125000]


I'll create a function to fetch the price associated with the kilometers on the vehcile

In [72]:
def get_price_mileage(km):
    price_km = {}
    for c in km["odometer_km"].unique():
        selected_rows = autos[autos["odometer_km"] == c]
        mean_price = selected_rows["price"].mean()
        price_km[c] = int(mean_price)
    return price_km

In [81]:
get_price_mileage(under_30)

{10000: 20550, 30000: 16608, 20000: 18448, 5000: 8873}

That's surprising, the cars with 5,000KM have the lowest mean price in the under 30,000KM category. Either the car is lemon or it has been in an accident. Check the proportion of cars with price equal to \\$5,000 and have been in a accident?

In [75]:
get_price_mileage(under_70)

{70000: 10927, 50000: 13812, 60000: 12385, 40000: 15499}

In [76]:
get_price_mileage(between80_100)

{80000: 9721, 90000: 8465, 100000: 8132}

In [82]:
get_price_mileage(above_100)

{150000: 3767, 125000: 6214}